teps/0046-triggertemplate-parameter-escaping.md

```diff
@@ -0,0 +1,295 @@
+   TriggerTemplates.
+
+### Non-Goals
+
+  * Providing for multiple forms of escaping data passing through
+    TriggerTemplates.
+
+### Use Cases (optional)
+ 
+## Requirements
+
+The basic requirement is that a JSON document including escaped quotes within a
+value should be able to rendered into a TriggerTemplate without generating
+invalid JSON.
+
+A common case where this might occur, is reverting a commit on GitHub, the
+generated title of the commit includes quotes for example:
+
+```json
+{
+  "action": "opened",
+  "number": 2,
+  "pull_request": {
+    "url": "https://api.github.com/repos/Codertocat/Hello-World/pulls/2",
+    "id": 279147437,
+    "title": "Revert \"Test Commit\"",
+  }
+}
+```
+
+the `pull_request.title` field is already escaped, so we need to pass this
+through unchanged, previously this was changed to:
+
+```json
+    "title": "Revert \""Test Commit\""",
+```
+
+Which is invalid JSON, and was causing errors.
+
+## Proposal
+
+This proposal seeks to add a field to TriggerTemplate ParamSpec structs, to
+allow a user to mark the field as "requiring to be escaped".
+
+At the point where the parameter's value is being inserted into the
+TriggerTemplate resourceTemplate, if the parameter is flagged for escaping, it
+should be inserted in such a way that the string won't break the subsequent
+parsing of the rendered template.
+
+### Notes/Caveats (optional)
+
+Ultimately, this is about escaping external data being received into the system,
+it does feel somewhat odd that the template insertion has to know whether or not
+the data should be escaped, clearly you could use a parameter in multiple places
+within a document, and in some of those places, escaping might be required or
+more appropriate than others, but it would be configured at the TriggerTemplate
+level, and not at the insertion level.
+
+### Risks and Mitigations
+
+We already have escaping, which can now be disabled, and the proposal is really
+just a fine-tuning of the functionality to allow users to control which strings
+they need to escape.
+
+### User Experience (optional)
+
+### Performance (optional)
+
+There's unlikely to be a significant performance hit from this approach, as the
+functionality already exists.
+
+## Design Details
+
+This would extend the [current](https://github.com/tektoncd/triggers/blob/master/pkg/apis/triggers/v1alpha1/param.go#L5-L15) `ParamSpec` struct.
+
+```
+type ParamSpec struct {
+    < existing fields omitted >>
+
+	// EscapeQuotes indicates whether or not the value should have double-quote
+	// symbols (") escaped before insertion into the templated bodies.
+	EscapeQuotes bool `json:"escapeQuotes,omitempty"`
+}
+```
+
+This would translate into the YAML as an addition to the declaration:
+
+```yaml
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: TriggerTemplate
+metadata:
+  name: escaped-param
+spec:
+  params:
+  - name: title
+    description: The title from the incoming body
+    escape: true
```

Comment:
If I understand correctly, in the example above, the title was already escaped `Revert \"Test Commit\"`, so to match that example we could have escape `false` here or refer to a different field perhaps?

The problem I see with this approach is that the escaping is controlled on the receiving side, which is not specific to the content producer, so using escaping in a template would bind the template to a specific content producer.

Trigger bindings are by nature specific to a content producer, so it feels to me that they would be a better place where to specify if escaping is needed or not.

Reply:
But at the same time, as well written by @bigkevmcd in https://github.com/tektoncd/community/pull/321/files#diff-9129936dcf33d43ca73c0a506952028c7f718c3e15d512ad586044d61579ad0bR174-R181, the receiving side also has context about the need (or not) for escaping. 

I like, among the alternatives proposed, the idea of controlling escaping when the variable associated to a parameter is used, for instance using an alternative syntax like `%(tt.params.name)` or some other syntax like `$(tt.params.name|escape)`.

Reply:
I suspect `%(tt.params.name)` is easier to implement `$(tt.params.name|escape)` implies (from Django/Jinja2) that the result of `tt.params.name` should be passed through a "filter" called escape?